src/main/java/com/google/devtools/build/lib/actions/FileContentsProxy.java

```diff
@@ -21,38 +21,60 @@
 
 /**
  * In case we can't get a fast digest from the filesystem, we store this metadata as a proxy to the
- * file contents. Currently it is a pair of a relevant timestamp and a "node id". On Linux the
- * former is the ctime and the latter is the inode number. We might want to add the device number in
+ * file contents. Currently it is two timestamps and a "node id". On Linux we
+ * use both ctime and mtime and inode number. We might want to add the device number in
  * the future.
  *
- * <p>For a Linux example of why mtime alone is insufficient, note that 'mv' preserves timestamps.
+ * <p>For a Linux example of ctime alone is insufficient, and we add node id,
+ * note that 'mv' preserves timestamps.
  * So if files 'a' and 'b' initially have the same timestamp, then we would think 'b' is unchanged
  * after the user executes `mv a b` between two builds.
+ *
+ * <p>On Linux we also need mtime for hardlinking sandbox, since updating the inode reference counter
+ * preserves mtime, but updates ctime. isModified() call can be used to compare two FileContentsProxys
+ * of hardlinked files.
  */
 public final class FileContentsProxy implements Serializable {
   private final long ctime;
+  private final long mtime;
   private final long nodeId;
 
+  /**
+   * Visible for serialization / deserialization. Do not use this method, but call {@link #create}
+   * instead.
+   */
+  public FileContentsProxy(long ctime, long mtime, long nodeId) {
+    this.ctime = ctime;
+    this.mtime = mtime;
+    this.nodeId = nodeId;
+  }
+
   /**
    * Visible for serialization / deserialization. Do not use this method, but call {@link #create}
    * instead.
    */
   public FileContentsProxy(long ctime, long nodeId) {
     this.ctime = ctime;
+    this.mtime = ctime;
```

Comment:
Don't know if mtime==ctime gives the right result, but not whether it matters either. Preferably I would get rid of this, but I don't know what the serialization requires.